Source/santa-driver/SantaDriverClient.cc

```diff
@@ -48,23 +48,21 @@ bool SantaDriverClient::start(IOService *provider) {
 }
 
 void SantaDriverClient::stop(IOService *provider) {
-  super::stop(provider);
   myProvider = nullptr;
   decisionManager = nullptr;
+  super::stop(provider);
 }
 
-IOReturn SantaDriverClient::clientClose() {
+IOReturn SantaDriverClient::clientDied() {
+  LOGI("Client died.");
   decisionManager->DisconnectClient(true);
-  return terminate(kIOServiceSynchronous) ? kIOReturnSuccess : kIOReturnError;
+  return terminate() ? kIOReturnSuccess : kIOReturnError;
 }
 
-bool SantaDriverClient::terminate(IOOptionBits options) {
-  decisionManager->DisconnectClient();
+IOReturn SantaDriverClient::clientClose() {
   LOGI("Client disconnected.");
-
-  if (myProvider && myProvider->isOpen(this)) myProvider->close(this);
-
-  return super::terminate(options);
+  decisionManager->DisconnectClient(false);
+  return terminate() ? kIOReturnSuccess : kIOReturnError;
```

Comment:
Is there any chance the client is still connected when calling terminate here? Should kIOServiceSynchronous or kIOServiceRequired be used?

Reply:
There shouldn't be, these methods are called in response to a client disconnecting and we only allow a single client.